### ISSUE TITLE
'Invalid HTTP response version

### ISSUE THREAD

**Issue by tvidal-34** [fcheslack/libZotero#16]
Dear all,

Since several weeks I can't update datas throught zotero API lib.

I updataed the libzotero to the v3 (last update this morning). But I have still a blank page with this error in my server log :

2014/12/12 10:37:58 [error] 10158#0: *3930918 FastCGI sent in stderr: "PHP message: PHP Fatal error:  Uncaught exception 'Zotero\Exception' with message 'Invalid HTTP response version: ' in /www/admin/build/libZoteroSingle.php:2990
Stack trace:
#0 /www/admin/build/libZoteroSingle.php(3483): Zotero\HttpResponse->__construct(false, Array, '', false, false)
#1 /www/admin/build/libZoteroSingle.php(2581): Zotero\HttpResponse::fromString(false)
#2 /www/admin/build/libZoteroSingle.php(1736): Zotero\Net->request(Array)
#3 /www/admin/build/libZoteroSingle.php(1896): Zotero\Library->request(Array)
#4 /www/admin/build/libZoteroSingle.php(2430): Zotero\Library->fetchItems(Array)
#5 /www/admin/majBD.php(34): Zotero\Library->loadItems(Array)
#6 /www/admin/majBD.php(1254): majBD->__construct()
#7 {main}
  thrown in /www/admin/build/libZoteroSingle.php on line 2990" while reading response header from upstream, client: x.x.x.x, server: www.toto.org, request: "GET /admin/majBD.php?debut=0 HTTP/1.1", upstream: "fastcgi://unix:/var/run/php5-fpm.sock:", host: "www.
totot.org", referrer: "http://www.toto.org/admin/index.php"

It's become to be urgent for us to be able to update our datas... 

Many thanks for your help.

Best regards
Thomas

**Comment by tvidal-34** [comment]
Dear fcheslack,

Sorry for the delay in my answer but I was on leave ! ...

First yes, the web site is behind a reverse proxy.... so I feel this can be the reason of this error...

Then I did what you wrote... and I have no more output in my log :

2015/01/12 15:16:17 [error] 2508#0: *177420 FastCGI sent in stderr: "PHP message: PHP Fatal error:  Uncaught exception 'Zotero\Exception' with message 'Invalid HTTP response version: ' in /www/www.toto.com/admin/build/libZoteroSingle.php:2991
Stack trace:
#0 /www/www.toto.com/admin/build/libZoteroSingle.php(3484): Zotero\HttpResponse->__construct(false, Array, '', false, false)
#1 /www/www.toto.com/admin/build/libZoteroSingle.php(2582): Zotero\HttpResponse::fromString(false)
#2 /www/www.toto.com/admin/build/libZoteroSingle.php(1736): Zotero\Net->request(Array)
#3 /www/www.toto.com/admin/build/libZoteroSingle.php(1896): Zotero\Library->request(Array)
#4 /www/www.toto.com/admin/build/libZoteroSingle.php(2430): Zotero\Library->fetchItems(Array)
#5 /www/www.toto.com/admin/majBD.php(34): Zotero\Library->loadItems(Array)
#6 /www/www.toto.com/admin/majBD.php(1254): majBD->__construct()
#7 {main}
  thrown in /www/www.toto.com/admin/build/libZoteroSingle.php on line 2991" while reading response header from upstream, client: x.x.x.x, server: www.toto.com, request: "GET /admin/majBD.php?debut=0 HTTP/1.1", upstream: "fastcgi://unix:/var/run/php5-fpm.sock:", host: "www.toto.com", referrer: "http://www.toto.com/admin/index.php" 

If you have other ideas..... many many thanks, this start to become critical for us !

Regards
Thomas

**Comment by tvidal-34** [comment]
Hi again,

The first fatal error comes from line 2991 :
if(!preg_match('|^\d.\d$|', $version)) { .....

What is the $version attends ?

Thomas

**Comment by emelin** [comment]
I obviously bumped into the same error but because of updating from osx 10.9 to 10.10, that is from apache2.2 to apache2.4.

Even tough the libZotero version used in my case might be an older version, I would appreciate a lot, if would could hunt down this error and make my use case of 'phpZoteroWebDAV' work again!

Here is my post to the respective repo. However, nothing change there since more than a year:

     https://github.com/ddean4040/phpZoteroWebDAV/issues/2

**Comment by tvidal-34** [comment]
Hi,

I made a copy of the application on another web server that use Apache2 and PHP5 (instead of Nginx and php5-fpm) and I have exactly the same error in my log :
[Wed Jan 21 15:27:47 2015] [error] [client 192.168.0.104] PHP Fatal error:  Uncaught exception 'Zotero\\Exception' with message 'Invalid HTTP response version: ' in /www/websites/www2.toto.org/admin/build/libZoteroSingle.php:2991\nStack trace:\n#0 /www/websites/www2.toto.org/admin/build/libZoteroSingle.php(3484): Zotero\\HttpResponse->__construct(false, Array, '', false, false)\n#1 /www/websites/www2.toto.org/admin/build/libZoteroSingle.php(2582): Zotero\\HttpResponse::fromString(NULL)\n#2 /www/websites/www2.toto.org/admin/build/libZoteroSingle.php(1736): Zotero\\Net->request(Array)\n#3 /www/websites/www2.toto.org/admin/build/libZoteroSingle.php(1896): Zotero\\Library->request(Array)\n#4 /www/websites/www2.toto.org/admin/build/libZoteroSingle.php(2430): Zotero\\Library->fetchItems(Array)\n#5 /www/websites/www2.toto.org/admin/majBD.php(34): Zotero\\Library->loadItems(Array)\n#6 /www/websites/www2.toto.org/admin/majBD.php(1254): majBD->__construct()\n#7 {main}\n  thrown in /www/websites/www2.toto.org/admin/build/libZoteroSingle.php on line 2991, referer: http://www2.toto.org/admin/

Please help !!! 

Many thanks

Thomas

**Comment by fcheslack** [comment]
did you turn libZotero debug logging back on after you updated? There should be more output than that regardless of errors.

**Comment by tvidal-34** [comment]
yes, I have still debug on as you suggested on 19th of December. And the output was what I put in the previous post..... :-(

**Comment by fcheslack** [comment]
Then I guess the debug is not making it to that error log. Maybe going to a different one, specific to that site? There should at least always be the "url being requested:" line which should be logged on every request.

**Comment by tvidal-34** [comment]
Hi,
I checked again both the PHP file and the error log output. I changed the php5-fpm setting to output all in php5-fpm.error.log and I have :

[21-Jan-2015 17:47:53] WARNING: [pool www] child 8518 said into stderr: "NOTICE: PHP message: PHP Notice:  Undefined index: HTTP_REFERER in /www/www.toto.org/banner.php on line 57"
[21-Jan-2015 17:47:55] WARNING: [pool www] child 8519 said into stderr: "NOTICE: PHP message: PHP Fatal error:  Uncaught exception 'Zotero\Exception' with message 'Invalid HTTP response version: ' in /www/www.toto.org/admin/build/libZoteroSingle.php:2991"
[21-Jan-2015 17:47:55] WARNING: [pool www] child 8519 said into stderr: "Stack trace:"
[21-Jan-2015 17:47:55] WARNING: [pool www] child 8519 said into stderr: "#0 /www/www.toto.org/admin/build/libZoteroSingle.php(3484): Zotero\HttpResponse->__construct(false, Array, '', false, false)"
[21-Jan-2015 17:47:55] WARNING: [pool www] child 8519 said into stderr: "#1 /www/www.toto.org/admin/build/libZoteroSingle.php(2582): Zotero\HttpResponse::fromString(false)"
[21-Jan-2015 17:47:55] WARNING: [pool www] child 8519 said into stderr: "#2 /www/www.toto.org/admin/build/libZoteroSingle.php(1736): Zotero\Net->request(Array)"
[21-Jan-2015 17:47:55] WARNING: [pool www] child 8519 said into stderr: "#3 /www/www.toto.org/admin/build/libZoteroSingle.php(1896): Zotero\Library->request(Array)"
[21-Jan-2015 17:47:55] WARNING: [pool www] child 8519 said into stderr: "#4 /www/www.toto.org/admin/build/libZoteroSingle.php(2430): Zotero\Library->fetchItems(Array)"
[21-Jan-2015 17:47:55] WARNING: [pool www] child 8519 said into stderr: "#5 /www/www.toto.org/admin/majBD.php(34): Zotero\Library->loadItems(Array)"
[21-Jan-2015 17:47:55] WARNING: [pool www] child 8519 said into stderr: "#6 /www/www.toto.org/admin/majBD.php(1254): majBD->__construct()"
[21-Jan-2015 17:47:55] WARNING: [pool www] child 8519 said into stderr: "#7 {main}"
[21-Jan-2015 17:47:55] WARNING: [pool www] child 8519 said into stderr: "  thrown in /www/www.toto.org/admin/build/libZoteroSingle.php on line 2991"

So nothing really better....

Many many thanks for your help.

Thomas

**Comment by fcheslack** [comment]
yeah, those are all still from php, not the script, which is just using error_log http://php.net/manual/en/function.error-log.php with defaults

**Comment by tvidal-34** [comment]
Hi,

As I said the error is coming from line 2991 about "Set the HTTP version".
I think that this waiting for something like HTTP/1.0
But I added just before the exception line :
error_log($version, 0);
And in my PHP log I have only a double quote (")

Thanks for your help

Thomas

**Comment by tvidal-34** [comment]
I set $version = "1.1"; 
And I have no more this error.
Now the error is coming on line 3511 :
[22-Jan-2015 10:08:02] WARNING: [pool www] child 10194 said into stderr: "NOTICE: PHP message: PHP Fatal error:  Uncaught exception 'Zotero\Exception' with message 'Unexpected content type not application/json. ' in /www/www.toto.org/admin/build/libZoteroSingle.php:3511"

Hope that can help....

Thomas

**Comment by fcheslack** [comment]
Those are notices and errors from the php (or server) process itself though. You still have no log output from libZotero itself. You could just echo or var_dump to the page if you can't figure out where the output from error_log is going on your setup.

That error is what I'd expect to happen if you set version manually. As I said earlier, it looks like the curl request is not actually getting any data, so everything will be an empty string when the response is parsed.

**Comment by tvidal-34** [comment]
Yes, I agree I didn't used a good way to test, and I also agree this comes because curl get nothing.
I just tested curl function by using example number 2 on this page : http://php.net/manual/fr/curl.examples.php
And I got "success".
Si no problem with my libcurl or its integration in php.

By default all error_log are going to /var/log/php5-fpm.log 
But I can use : 
error_log($responseBody, 3, "/tmp/zotero.log");
Is it what you want me to do ?

Thanks again.

Thomas

**Comment by fcheslack** [comment]
Yes, if you use that for the libZoteroDebug function you should be able to see the urls being requested, and you can use a separate test script to test if curl can retrieve those urls or not.

**Comment by tvidal-34** [comment]
ok, I will do it next monday... I'll keep you informed. Many thanks.

**Comment by emelin** [comment]
@tvidal-34 any good (or bad) news about that issue yet?

**Comment by tvidal-34** [comment]
@emelin : new post here : https://github.com/fcheslack/libZotero/issues/18

**Comment by emelin** [comment]
@tvidal-34 thank you for pointing me there! I downloaded the ssl certificate once more into Firefox (even though I am sure to have done this before) and now my usage of the (older) libZotero that is part of 'phpZoteroWebDAV' seems to work again! My current sync run will take ages to handle all the Gigs. However, if anything further goes wrong, be sure I'll be back and report. Long live this API!

**Comment by tvidal-34** [comment]
Dear Emelin,
Happy that help you !!
I did the same as you a week ago, but without the same result. Can you describe more in depth where and how you add the certificate path for Curl ?
All the best

Thomas

**Comment by emelin** [comment]
@tvidal-34 Yes happiness here, because the sole and simple thing that I used to run since years works again on Appache 2.4 and OSX 10.10.2. This is 'https://github.com/ddean4040/phpZoteroWebDAV' which uses an older version of the present API. Probably disappointing for your, I have not fiddled around with curl at all. The only thing I did, was accepting a security exception of my own certificates in Firefox (which I surely did sometime before as the problem above initially hit me). It may well be that something at the Zotero side has changed meanwhile. In this blur, my happiness remains kind of magical.
